### PR TITLE
fix: lease keys in conflict pool

### PR DESCRIPTION
**Pull request by bsbds**
Depends-On: #852 

Fix lease related logic in conflict pools.

Please briefly answer these questions:

* what problem are you trying to solve? (or if there's no problem, what's the motivation for this change?)

* what changes does this pull request make?

* are there any non-obvious implications of these changes? (does it break compatibility with previous versions, etc)

### CODE REVIEW

File: crates/xline/src/conflict/mod.rs [hunk 1]
```diff
@@ -1,15 +1,23 @@
...
+};
+
+use crate::storage::lease_store::{Lease, LeaseCollection};
 
 use self::{
     spec_pool::{ExclusiveSpecPool, KvSpecPool, LeaseSpecPool},
     uncommitted_pool::{ExclusiveUncomPool, KvUncomPool, LeaseUncomPool},
 };
-
+// TODO: Refine code to improve reusability for different conflict pool types
```
Comment (Phoenix500526) [inline crates/xline/src/conflict/mod.rs:20]: What does different conflict pool types mean here？ Is there any example?

File: crates/xline/src/conflict/tests.rs [hunk 2]
```diff
@@ -181,14 +184,229 @@ fn exclusive_ucp_operations_are_ok() {
...
+The following test cases verify that insert and remove operations in a
+conflict pool are independent of any external state.
+
+Specifically, we need to query `LeaseCollection` for kv and lease requests
+and the `LeaseCollection` might be mutated sometime between a insert and
+a remove, potentially leading to an inconsist state in our conflict pool.
+*/
+
+#[test]
+fn kv_sp_mutation_no_side_effect() {
```
Comment (Phoenix500526) [inline crates/xline/src/conflict/tests.rs:331]: Please leave a comment to describe what the side effect mean here.

File: crates/xline/src/conflict/tests.rs [hunk 3]
```diff
@@ -181,14 +184,229 @@ fn exclusive_ucp_operations_are_ok() {
...
+    lease_collection
+        .attach(1, "foo".as_bytes().to_vec())
+        .unwrap();
+    let kv_put = gen.gen_put("foo");
+    let lease_revoke = gen.gen_lease_revoke(1);
+
+    sp.insert_if_not_conflict(lease_revoke.clone());
+    assert!(sp.insert_if_not_conflict(kv_put.clone()).is_some());
+    lease_collection.detach(1, "foo".as_bytes()).unwrap();
+    sp.remove(&lease_revoke);
```
Comment (Phoenix500526) [inline crates/xline/src/conflict/tests.rs:346]: What does this assertion prove? If you remove the lease_revoke from the sp, of course you won't get conflict while inserting the kv_put request.

File: crates/xline/src/conflict/tests.rs [hunk 4]
```diff
@@ -181,14 +184,229 @@ fn exclusive_ucp_operations_are_ok() {
...
+        .attach(1, "foo".as_bytes().to_vec())
+        .unwrap();
+    let kv_put = gen.gen_put("foo");
+    let lease_revoke = gen.gen_lease_revoke(1);
+
+    sp.insert_if_not_conflict(kv_put.clone());
+    assert!(sp.insert_if_not_conflict(lease_revoke.clone()).is_some());
+    lease_collection.detach(1, "foo".as_bytes()).unwrap();
+    sp.remove(&kv_put);
+    assert!(sp.insert_if_not_conflict(kv_put).is_none());
```
Comment (Phoenix500526) [inline crates/xline/src/conflict/tests.rs:367]: Same as above

**Review by Phoenix500526 — COMMENTED** [review]
I've noticed that this PR contains changes to xline-proto, but I don't see any PR in the xline-proto repo. Did you forget to create a PR for xline-proto?

**Comment by bsbds** [comment]
> I've noticed that this PR contains changes to xline-proto, but I don't see any PR in the xline-proto repo. Did you forget to create a PR for xline-proto?

Opened in https://github.com/xline-kv/xline-proto/pull/12

File: crates/xline/src/conflict/tests.rs (hunk 2 above)
Comment (bsbds) [reply crates/xline/src/conflict/tests.rs:331]: Explained in L321 - L328

File: crates/xline/src/conflict/mod.rs (hunk 1 above)
Comment (bsbds) [reply crates/xline/src/conflict/mod.rs:20]: For instance, `KvSpecPool` `LeaseSpecPool` `KvUncomPool` and`LeaseUncomPool` might have some redundant code.

File: crates/xline/src/conflict/tests.rs (hunk 3 above)
Comment (bsbds) [reply crates/xline/src/conflict/tests.rs:346]: It aims to ensure that the `lease_revoke` command is correctly removed.

**Review by Phoenix500526 — CHANGES_REQUESTED** [review]
LGTM，please squash these commits into one

**Review by Phoenix500526 — APPROVED** [review]
LGTM

**Pull request merged**